src/signer_node/mod.rs

```diff
@@ -131,120 +131,145 @@ impl<T: TapyrusApi, C: ConnectionManager> SignerNode<T, C> {
-        let closure = move |message: Message| match sender.send(message) {
-            Ok(_) => ControlFlow::Continue,
-            Err(error) => {
-                log::warn!("Happened error!: {:?}", error);
-                ControlFlow::Break(())
-            }
-        };
-        let id = self.params.signer_id;
-        let _handler = self.connection_manager.start(closure, id);
-
-        log::info!("Start Key generation Protocol");
-        // Idle 5s, before node starts Key Generation Protocol communication.
-        // To avoid that nodes which is late to startup can't receive messages.
-        log::info!("Idle 5 secs... ");
-        std::thread::sleep(Duration::from_secs(5));
-
-        // Start First Round
-        log::info!("Start block creation rounds.");
-        self.start_next_round();
-
-        // get error_handler that is for catch error within connection_manager.
-        let connection_manager_error_handler = self.connection_manager.error_handler();
         loop {
-            // After process when received message. Get message from receiver,
-            // then change that state in main thread side.
-            // messageを受け取った後の処理。receiverからmessageを受け取り、
-            // stateの変更はmain thread側で行う。
-            match &self.stop_signal {
-                Some(ref r) => match r.try_recv() {
-                    Ok(_) => {
-                        log::warn!("Stop by Terminate Signal.");
-                        self.round_timer.stop();
-                        break;
-                    }
-                    Err(std::sync::mpsc::TryRecvError::Empty) => {
-                        // Stop signal is empty. Continue to run. Do nothing.
-                    }
-                    Err(e) => {
-                        panic!("{:?}", e);
-                    }
-                },
-                None => {
-                    // Stop signal receiver is not set. Do nothing.
+            log::info!("Start thread for redis subscription");
+            let (sender, receiver): (Sender<Message>, Receiver<Message>) = channel();
+            let closure = move |message: Message| match sender.send(message) {
+                Ok(_) => ControlFlow::Continue,
+                Err(error) => {
+                    log::warn!("Happened error!: {:?}", error);
+                    ControlFlow::Break(())
+                }
+            };
+            let id = self.params.signer_id;
+            let handler = self.connection_manager.start(closure, id);
+
+            log::info!("Start Key generation Protocol");
+            // Idle 5s, before node starts Key Generation Protocol communication.
+            // To avoid that nodes which is late to startup can't receive messages.
+            log::info!("Idle 5 secs... ");
+            std::thread::sleep(Duration::from_secs(5));
+
+            // Start First Round
+            log::info!("Start block creation rounds.");
+            self.start_next_round();
```

Comment:
redis が停止した場合、いずれ全員が master としてラウンドを開始して、round_duration の時間だけ sleep に入る挙動をしそうな気がします。そうなると、redisが復旧してもスムーズにネットワークが復旧しないと思います。

この行よりも前で `handle_connection` して、接続エラーの場合は少し sleep してredisの復旧を待つ挙動にしたらどうかと思いました。

Reply:
Calling handle_connection before self.start_next_round() returns None which means no error.
This is because ConnectionManager#start calls get_connection but it raises an error in another thread.
So I add the test_connection() which checks the connection in this thread.

Reply:
fb56c96